plasmapy/particles/ionization_state.py

```diff
@@ -2,30 +2,169 @@
+            #       This inconsistency should be fixed.
+
+            ionic_fraction_within_tolerance = np.isclose(
+                self.ionic_fraction, other.ionic_fraction, rtol=1e-15,
+            )
+
+            number_density_within_tolerance = u.isclose(
+                self.number_density, other.number_density, rtol=1e-15,
+            )
+
+            return all(
+                [ionic_fraction_within_tolerance, number_density_within_tolerance,]
+            )
+
+        except Exception as exc:
+            raise TypeError(
+                "Unable to ascertain equality between the following objects:\n"
+                f"  {self}"
+                f"  {other}"
+            ) from exc
+
+    @particle_input
+    def __init__(self, ion: Particle, ionic_fraction=None, number_density=None):
+        self._particle = ion
+        self.ionic_fraction = ionic_fraction
+        self.number_density = number_density
+
+    def __repr__(self):
+        return (
+            f"IonicFraction({repr(self.ionic_symbol)}, "
+            f"ionic_fraction={self.ionic_fraction})"
+        )
+
+    @property
+    def ionic_symbol(self) -> str:
+        """The symbol of the ion."""
+        return self._particle.ionic_symbol
+
+    @property
+    def integer_charge(self) -> Integral:
+        """The integer charge of the ion."""
+        return self._particle.integer_charge
+
+    @property
+    def ionic_fraction(self) -> Real:
+        r"""
+        The fraction of particles of an element that are at this
+        ionization level.
+
+        Notes
+        -----
+        An ionic fraction must be in the interval :math:`[0, 1]`.
+
+        If no ionic fraction is specified, then this attribute will be
+        assigned the value of `~numpy.nan`.
+        """
+        return self._ionic_fraction
+
+    @ionic_fraction.setter
+    def ionic_fraction(self, ionfrac: Optional[Real]):
+        if ionfrac is None or np.isnan(ionfrac):
+            self._ionic_fraction = np.nan
+        else:
+            try:
+                out_of_range = ionfrac < 0 or ionfrac > 1
+            except TypeError:
+                raise TypeError(f"Invalid ionic fraction: {ionfrac}")
+            else:
+                if out_of_range:
+                    raise ValueError(f"The ionic fraction must be between 0 and 1.")
+                else:
+                    self._ionic_fraction = ionfrac
+
+    @property
+    def number_density(self) -> u.m ** -3:
+        """The number density of the ion."""
+        return self._number_density.to(u.m ** -3)
+
+    @number_density.setter
+    @validate_quantities(
+        n={
+            "units": u.m ** -3,
+            "can_be_negative": False,
+            "can_be_inf": False,
+            "none_shall_pass": True,
+        },
+    )
+    def number_density(self, n: u.m ** -3):
+        if n is None:
+            self._number_density = np.nan * u.m ** -3
+        else:
+            self._number_density = n.to(u.m ** -3)
```

Comment:
I think I'm going to keep a bookmark on this to remember how to use `validate` on properties :smile:

Reply:
Turns out, trying to figure this out was the inspiration for #797!